tools/lsp/server_loop.rs

```diff
@@ -609,6 +678,66 @@ pub mod tests {
         (dc, url, diag)
     }
 
+    pub fn complex_document_cache(
```

Comment:
This module is becoming big and is being accessed from a different module as well.  I would say it could be made into its onw file.  (Optional for this PR)

Reply:
Agreed, that will be much nicer!

Done in the next push.